### PR TITLE
ldap query minor refactoring

### PR DESCRIPTION
**Pull request by ronfroy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes 
| License       | MIT

### CODE REVIEW

**Review by guilliamxavier — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php [hunk 1]
```diff
@@ -62,20 +63,18 @@ public function execute()
...
-                    break;
-                default:
-                    throw new LdapException(sprintf('Could not search in scope "%s".', $this->options['scope']));
+            $funcMap = array(
+                QueryInterface::SCOPE_BASE => 'ldap_read',
+                QueryInterface::SCOPE_ONE => 'ldap_list',
+                QueryInterface::SCOPE_SUB => 'ldap_search',
+            );
+
+            if (\in_array($this->options['scope'], $funcMap, true)) {
```
[the remark this answers is not in the document]
Comment (ronfroy) [reply src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php]: @guilliamxavier this is fixed

[the remark this answers is not in the document]
Comment (guilliamxavier) [reply src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php]: Thanks. I didn't know that carsonbot would change the status, I will now try to reset it:

Status: Needs Review

**Comment by guilliamxavier** [comment]
Status: Needs Review

File: src/Symfony/Component/Ldap/Adapter/AbstractQuery.php [hunk 2]
```diff
@@ -32,12 +32,12 @@ public function __construct(ConnectionInterface $connection, string $dn, string
             'maxItems' => 0,
             'sizeLimit' => 0,
             'timeout' => 0,
-            'deref' => static::DEREF_NEVER,
+            'deref' => QueryInterface::DEREF_NEVER,
```
Comment (iltar) [inline src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: This will break backwards compatibility, as you could use them in a child implementation and overwrite them. @csarrazi do you know if that was intentional?

**Review by linaori — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Ldap/Adapter/AbstractQuery.php (hunk 2 above)
Comment (ronfroy) [reply src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: This is, but it's a design issue too.

Comment (ronfroy) [reply src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: @iltar I can not make it deprecated and fixed it in the same time. But i dont think this is ever used.

Comment (iltar) [reply src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: You can trigger the deprecation, which should cause developers to fix the issue. It'll have to be fixed in 5.0. It's up to the core members to decide whether or not this is an acceptable BC break

Comment (csarrazi) [reply src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: Actually, one of the things I'm unsatisfied with is the fact that the values in QueryInterface are actually extracted from the `ext_ldap` implementation, which is actually problematic as this means that the implementation leaks into the interface.

This was actually the rationale behind using `static::` instead of QueryInterface, as the goal was that other implementations could use other values.

For 5.x, the actual values used in the QueryInterface should instead be defined in the actual implementation instead of the interface, and the interface constants should rather be string values.

Comment (csarrazi) [reply src/Symfony/Component/Ldap/Adapter/AbstractQuery.php]: I agree. We should create a `LdapOption`, which would contain string values for each constant.
The actual value (used by `ext_ldap` should be limited to the actual implementation, so should only be present in the query implementation.

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php [hunk 3]
```diff
@@ -37,7 +38,7 @@ public function __destruct()
         $con = $this->connection->getResource();
         $this->connection = null;
 
-        if (null === $this->search || false === $this->search) {
+        if (\in_array($this->search, array(null, false), true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php]: please revert this change, it's unrelated and not needed

File: src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php [hunk 4]
```diff
@@ -62,20 +63,18 @@ public function execute()
...
-                    break;
-                default:
-                    throw new LdapException(sprintf('Could not search in scope "%s".', $this->options['scope']));
+            $funcMap = array(
+                QueryInterface::SCOPE_BASE => 'ldap_read',
+                QueryInterface::SCOPE_ONE => 'ldap_list',
+                QueryInterface::SCOPE_SUB => 'ldap_search',
+            );
+
+            if (!\array_key_exists($this->options['scope'], $funcMap)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Ldap/Adapter/ExtLdap/Query.php]: "isset" would do it

**Comment by ronfroy** [comment]
@nicolas-grekas reverted

**Comment by fabpot** [comment]
As this was "just" a refactoring and because it breaks BC, shall we close?

**Comment by csarrazi** [comment]
Agreed
